### ISSUE TITLE
Enable user to provide jacobian for nonlinear constraints 

### ISSUE THREAD

**Issue by Osburg** [basf/doe#14]
If no jacobian is provided for a constraint, IPOPT will approximate it using finite differences. For linear constraints the jacobian is generated internally by doe during the processing of the provided `opti.Problem`. For arbitrary nonlinear constraints this cannot be done easily. However, in high dimensions finite the difference approximation is too time consuming.
Therefore, there should be an option for the user to provide a jacobian for nonlinear constraints.

To illustrate the problem: Here we have 10 variables with 1 nonlinear constraint and a linear model.
```
problem = opti.Problem(
    inputs=opti.Parameters([opti.Continuous(f"x{i+1}", [0, 1]) for i in range(10)]),
    outputs=[opti.Continuous("y")],
    constraints=[opti.NonlinearEquality("(x1**2 + x2**2 + x3**2 + x4**2 + x5**2 + x6**2 + x7**2 + x8**2 + x9**2)**0.5 - x10")],
)

result = find_local_max_ipopt(problem, "linear", ipopt_options={"disp":5, "maxiter": 100})
result.round(3)
```
And now the same with a linear constraint (mixture constraint)
```
problem = opti.Problem(
    inputs=opti.Parameters([opti.Continuous(f"x{i+1}", [0, 1]) for i in range(10)]),
    outputs=[opti.Continuous("y")],
    constraints=[opti.LinearEquality(names=[f"x{i+1}" for i in range(10)], rhs=1)],
)

result = find_local_max_ipopt(problem, "linear", ipopt_options={"disp":5, "maxiter": 100})
result.round(3)
```

When you run both examples you will see that even in case of a moderate problem size of 10 variables with only one nonlinear constraint we observe that the optimization slows down significantly (and maybe even get worse results due to inaccuracies of the finite difference approximation).

**Comment by Osburg** [comment]
@DavidWalz What do you think about adding jacobians as an (optional) attribute for opti constraints as suggested in [this issue](https://github.com/basf/mopti/issues/39)?
If you support this idea I would use it as possibility for the user to manually provide constraint jacobians.

**Comment by DavidWalz** [comment]
Good point. And in addition to the figure differences the evaluations are also showed down by the repeated pandas conversions.

As correctly writing out the derivative gets complicated quite easily, this calls for autodif, which would also be beneficial for custom models. Additionally using Mindtpy from pyomo to handle integer variables would be great.

@dlinzner-bcs has ported the main doe functionality to https://github.com/experimental-design/bofire
with the same limitations. 

Long story short. I think a switch of solver and autodif is required to bring d-optimal design to the next level and this involves a major overhaul.

**Comment by DavidWalz** [comment]
Interesting findings! Can you share your code for this analysis? If this is the case then it would be indeed worthwhile to already tackle the gradients for the nonlinear part separately from a possible overhaul (there is no timeline). Feel free to open a PR in bofire.

**Comment by Osburg** [comment]
Yes, you can just run `doe/examples/runtime.ipynb` in [this branch](https://github.com/Osburg/doe/tree/runtime) of my fork. Ok, will create a PR in bofire with support for nonlinear constraints and the optional constraint gradients and names attribute for constraints as soon as I have time for it.
